### PR TITLE
adds an event bus policy resource

### LINKED ISSUE #16838: AWS EventBridge bus resource policy
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS have added support for resource policies that allow more granular control over who can subscribe or publish to a bus. 

They've chosen to expose this behaviour by extending the existing PutPermission action on the eventbridge resource. This API endpoint now accepts _either_ a policy document (the new behaviour), or an action, statement, principal and condition that AWS will merge with the existing resource policy (previous behaviour). Each `permission` resource effectively becomes a new statement on the policy.

The `aws_cloudwatch_event_permission` resource is quite constrained, and supports the second, older behaviour. When loading state, the resource parses the policy document and matches statement IDs to figure out which `permission` resources are present. I'm unable to use this resource as is, because it doesn't support the actions or condition keys that I want to use, but which are supported by the policy language.

If I create the policy manually on my event bus, Terraform is unable to load the bus as a `permission` resource because it doesn't recognise the actions I've added.

I would like to submit a new `aws_cloudwatch_event_bus_policy` resource that allows control of the resource policy document directly. This would not be compatible with the existing `permission` resource. Policies created directly in EventBridge might not be readable as `aws_cloudwatch_event_permission` resources. I suspect that having direct access to the resource policy means that the `aws_cloudwatch_event_permission` resource can be fully deprecated.

### New or Affected Resource(s)

* New: aws_cloudwatch_event_bus_policy
* Replace: aws_cloudwatch_event_permission

### Potential Terraform Configuration

```hcl

resource "aws_cloudwatch_event_bus_policy" "test" {
   event_bus_name = "my-excellent-event-bus"
   policy = some_policy_document.json
}

```

### References

* [Resource policy announcement for EventBridge](https://aws.amazon.com/blogs/compute/simplifying-cross-account-access-with-amazon-eventbridge-resource-policies/) 
* [Syntax for the PutPermission call](https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_PutPermission.html#API_PutPermission_RequestSyntax) which supports _either_ a policy or the components of a statement.

### PR DESCRIPTION
**Pull request by cohen990**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #16838

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```
Adds support for resource policies on cloudwatch eventbridge

```

Output from acceptance testing:

```

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=CloudwatchEventBusPolicy -timeout 120m
=== RUN   TestAccAWSCloudwatchEventBusPolicy_basic
=== PAUSE TestAccAWSCloudwatchEventBusPolicy_basic
=== CONT  TestAccAWSCloudwatchEventBusPolicy_basic
--- PASS: TestAccAWSCloudwatchEventBusPolicy_basic (19.01s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	24.990s

```

### CODE REVIEW

**Comment by cohen990** [comment]
Hi, I was hoping to get some guidance on why the acceptance test is failing. I can't figure out where the whitespace difference may have been introduced nor can I figure out where the plan is actually being run.

From what I can tell, I'm storing everything correctly into the state.

**Comment by cohen990** [comment]
Managed to get the diff suppression working!

**Comment by heitorlessa** [comment]
Thanks a lot for the contribution @cohen990 - This is the last feature to complete EventBridge feature parity with CloudFormation. I'll look into it early next week.

At a first glance, it looks straightforward enough, so might be able to finish triaging next week the latest

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy.go:34]: This was updated in latest releases to also take into account Partner buses that slightly differ in ARN.

```suggestion
				ValidateFunc: validateCloudWatchEventBusNameOrARN,
```

Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy.go:34]: We changed in recent releases to account for partner buses that differ in ARN

```suggestion
				ValidateFunc: validateCloudWatchEventBusNameOrARN,
```

File: aws/resource_aws_cloudwatch_event_bus_policy_test.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy_test.go:28]: Could you add an extra step to test for updates? 

e.g. https://github.com/hashicorp/terraform-provider-aws/blob/7247dd06a08d9647ac061d387e5a185ae948efe6/aws/resource_aws_iam_group_policy_test.go#L40

File: aws/resource_aws_cloudwatch_event_bus_policy_test.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy_test.go:25]: Could you check for properties in the policy to confirm they match key details like principal and statement sid?

e.g. `resource.TestCheckResourceAttr(resourceName1, "principal", principal),`

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy.go:64]: I might be wrong, but I'm on the fence about this not being a combination of eventBus + policy statement ID (unique), like in the `event_permission` resource.

I'd defer to core maintainers on this one

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy.go:159]: Could you create a test for this? e.g. https://github.com/hashicorp/terraform-provider-aws/blob/7247dd06a08d9647ac061d387e5a185ae948efe6/aws/resource_aws_cloudwatch_event_permission_test.go#L294

File: website/docs/r/cloudwatch_event_bus_policy.html.markdown
Comment (heitorlessa) [inline website/docs/r/cloudwatch_event_bus_policy.html.markdown:23]: Could you add this resource for a more complete example? e.g. a policy with a Principal using a dummy AWS Account number

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy.go:34]: This has been updated to account for partner bus ARN that differs slightly
```suggestion
				ValidateFunc: validateCloudWatchEventBusNameOrARN,
```

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy.go:135]: Could you create a test updating a policy?

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy.go:159]: Could you create a test for deletion?

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy.go:64]: I'm on the fence of this not being a combination of bus name + statement Id - I suspect it's okay given you can have a single resource policy per bus, so I'll defer to core maintainers on that.

File: aws/resource_aws_cloudwatch_event_bus_policy_test.go
Comment (heitorlessa) [inline aws/resource_aws_cloudwatch_event_bus_policy_test.go:27]: Could you check for policy properties like statement_id and principal? e.g. check whether policy matches what's being created

File: website/docs/r/cloudwatch_event_bus_policy.html.markdown
Comment (heitorlessa) [inline website/docs/r/cloudwatch_event_bus_policy.html.markdown:23]: Could you add this sample policy as part of the example to improve docs experience?

**Review by heitorlessa — CHANGES_REQUESTED** [review]
Looks mostly good - missing tests on update/delete, and check on policy properties for safety update.

We also updated validateFunc since this PR was created so you can simply accept the suggestion as tests are broken now.

After these happy to send to prioritization so we can complete EventBridge feature coverage on TF

Thanks a lot for the hard work on this!

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (FrancescoFucile-CAZ) [reply aws/resource_aws_cloudwatch_event_bus_policy.go:159]: Are you referring to a test for "resourceAwsCloudWatchEventBusPolicyDelete"? Something like deleting the policy and verifying it doesn't exist anymore using the SDK?

**Comment by heitorlessa** [comment]
Yes, that’s correct.

File: aws/resource_aws_cloudwatch_event_bus_policy_test.go
Comment (FrancescoFucile-CAZ) [reply aws/resource_aws_cloudwatch_event_bus_policy_test.go:25]: The problem with this check is that we are setting the whole policy JSON text as a field in the "event_bus_policy" resource configuration. We are therefore not able to check fields individually, we can just compare the JSON document in the state with a reference. Since this JSON is generated at runtime we are not able to set such reference as a parameter for the test function.
We found a similar scenario for the "[aws_iam_role](https://github.com/Cazoo-uk/terraform-provider-aws/blob/42c30754145d2a5a1bfb2a3eb5e765a5e9caa782/aws/data_source_aws_iam_role_test.go#L70)". In this case the tests make use of a data source for re-importing the configuration of the generated resource into the state and compare its fields with the ones of the previously generated resource. Implementing a solution like this would require us to implement the "event_bus_policy" data source.
Is there a different way to handle this case?

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (FrancescoFucile-CAZ) [reply aws/resource_aws_cloudwatch_event_bus_policy.go:64]: Considering the "aws_cloudwatch_event_bus_policy" resource supports multiple statements we think that making it unique for each event bus won't introduce limitations. Also this implementation appears to be consistent with the SDK/CLI, which don't model event bus permissions as independent resources.

**Comment by heitorlessa** [comment]
Thanks for the comments, Francesco! I'll have a second look tomorrow,
especially the deletion test.

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (heitorlessa) [reply aws/resource_aws_cloudwatch_event_bus_policy.go:64]: I don't disagree. I don't just understand what the rationale was behind `aws_cloudwatch_event_permission` to create an unique ID back then: https://github.com/hashicorp/terraform-provider-aws/blob/fafcf78238182983aa8a5238c326f9049448aa70/aws/resource_aws_cloudwatch_event_permission.go#L103-L102

Comment (FrancescoFucile-CAZ) [reply aws/resource_aws_cloudwatch_event_bus_policy.go:64]: The "aws_cloudwatch_event_permission" resource only defines one statement in the policy, and it's therefore identified by it. The resource "aws_cloudwatch_event_policy" defines the complete policy for the event bus (possibly with multiple statements) and it's intended to be unique for each event bus.

**Comment by heitorlessa** [comment]
Not that I’m aware no. I’ve tried looking for references to the
Organization’s resource test where we use an extra data caller identity
which can be tested using TestResourceAttrPair, but it wouldn’t in this
case.

My best guess would be to create a function to JSON decode and assert
whether a given key has the expected value - this would be useful to test
the Update logic that isn’t tested atm, for example.

**Comment by cohen990** [comment]
@heitorlessa this should be ready for another review :)

File: aws/resource_aws_cloudwatch_event_bus_policy_test.go
Comment (heitorlessa) [reply aws/resource_aws_cloudwatch_event_bus_policy_test.go:38]: This now does a deep diff from TF state and what's available in the bus. It does the job and run through the entire logic IMHO and state.

https://github.com/hashicorp/terraform-provider-aws/pull/16874/files?file-filters%5B%5D=.go#diff-23812e233d355579267c55e833a81e4aa3a286298ee1477b978cbcd22ec7ad0cR141

Deferring to Hashicorp core team if they want to be nitpick and check whether X key has changed.

File: aws/resource_aws_cloudwatch_event_bus_policy_test.go
Comment (heitorlessa) [reply aws/resource_aws_cloudwatch_event_bus_policy_test.go:39]: `testAccAWSCloudwatchEventBusPolicyConfigUpdate` added.

File: aws/resource_aws_cloudwatch_event_bus_policy_test.go
Comment (heitorlessa) [reply aws/resource_aws_cloudwatch_event_bus_policy_test.go:28]: https://github.com/hashicorp/terraform-provider-aws/pull/16874/files?file-filters%5B%5D=.go#diff-23812e233d355579267c55e833a81e4aa3a286298ee1477b978cbcd22ec7ad0cR141

Checks all of it, so we're good here.

File: aws/resource_aws_cloudwatch_event_bus_policy.go
Comment (heitorlessa) [reply aws/resource_aws_cloudwatch_event_bus_policy.go:135]: `testAccAWSCloudwatchEventBusPolicyConfigUpdate` added.

**Review by heitorlessa — APPROVED** [review]
LGTM - Hashicorp core team might want to check whether it needs a attribute check on update test or whether the current deep value check of what's in TF state vs what's available in the Cloud suffice.

Sending to Hashicorp folks to prioritize it ;) Thanks a lot for those changes @cohen990 !

---

**Latest test results**

```go
make testacc TESTARGS='-run=TestAccAWSCloudwatchEventBusPolicy'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudwatchEventBusPolicy -timeout 180m
=== RUN   TestAccAWSCloudwatchEventBusPolicy_basic
=== PAUSE TestAccAWSCloudwatchEventBusPolicy_basic
=== RUN   TestAccAWSCloudwatchEventBusPolicy_disappears
=== PAUSE TestAccAWSCloudwatchEventBusPolicy_disappears
=== CONT  TestAccAWSCloudwatchEventBusPolicy_basic
=== CONT  TestAccAWSCloudwatchEventBusPolicy_disappears
--- PASS: TestAccAWSCloudwatchEventBusPolicy_basic (51.69s)
--- PASS: TestAccAWSCloudwatchEventBusPolicy_disappears (146.29s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       149.018s

```

**Comment by heitorlessa** [comment]
Actually, @cohen990, it seems that the changelog vanished in the last merge - Could you re-add it? If it wasn't there in the first place and I dreamed it :D, here's what the file should look like so CI is happy about it.

**File**: `./changelog/16874.txt`

Content: 

```release-note:new-resource
aws_cloudwatch_event_bus_policy
```

File: website/docs/r/cloudwatch_event_bus_policy.html.markdown
Comment (heitorlessa) [reply website/docs/r/cloudwatch_event_bus_policy.html.markdown]: docs added with multiple examples. LGTM

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSCloudwatchEventBusPolicy_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudwatchEventBusPolicy_ -timeout 180m
=== RUN   TestAccAWSCloudwatchEventBusPolicy_basic
=== PAUSE TestAccAWSCloudwatchEventBusPolicy_basic
=== RUN   TestAccAWSCloudwatchEventBusPolicy_disappears
=== PAUSE TestAccAWSCloudwatchEventBusPolicy_disappears
=== CONT  TestAccAWSCloudwatchEventBusPolicy_basic
=== CONT  TestAccAWSCloudwatchEventBusPolicy_disappears
--- PASS: TestAccAWSCloudwatchEventBusPolicy_basic (25.64s)
--- PASS: TestAccAWSCloudwatchEventBusPolicy_disappears (133.16s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	136.188s
```

##### GovCloud

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSCloudwatchEventBusPolicy_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudwatchEventBusPolicy_ -timeout 180m
=== RUN   TestAccAWSCloudwatchEventBusPolicy_basic
=== PAUSE TestAccAWSCloudwatchEventBusPolicy_basic
=== RUN   TestAccAWSCloudwatchEventBusPolicy_disappears
=== PAUSE TestAccAWSCloudwatchEventBusPolicy_disappears
=== CONT  TestAccAWSCloudwatchEventBusPolicy_basic
=== CONT  TestAccAWSCloudwatchEventBusPolicy_disappears
--- PASS: TestAccAWSCloudwatchEventBusPolicy_basic (35.24s)
--- PASS: TestAccAWSCloudwatchEventBusPolicy_disappears (138.06s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	141.106s
```

**Pull request merged**
